don't close collection on sync (performance)

Arthur-Milchior opened the pull request:
In https://github.com/ankidroid/Anki-Android/pull/5972 we considered the question: why does full sync require a model loading ?

This led me to this PR. It's a small but complex one.

In upstream, the collection is closed on sync only for full download. Of course, since the collection is replaced by an entirely new file. Actually, it's not even closed for exception, which I didn't expect.

On the other hand, currently, our code ALWAYS close the collection. Which is entirely useless. On full upload, it reopens the collection immediatly to check the content, optimize the DB, etc... before uploading. This is a huge loss of time, as we must read a lot of data again from the DB.

It even close the collection on normal sync, when there are no problem. I doubt there is any reason to do it because:
1. I use this code in my "merge" branch, and everything seems to work
2. the Syncer class deals with all change received from the server, so we don't need to re read the database content to find the changes. And anyway, we will only see the changes that were put in it before the data base closed.

Note that I don't even need to close the collection on full download here, since the syncer takes care of it. Instead, I just need to close it on exception, in order to avoid having even more exceptions.

This is clearly a time saver, as loading the collection is a costly process


# How it has been tested

Making change on my computer. Syncing them. Syncing my phone. Seeing the change on my phone. Making chaneg, syncing my phone, syncing the computer, seeing change on the computer.

Doing change on the smartphone, doing a full download, seeing that those change are cancelled and we're back to last version.
Doing change on phone, doing full upload. Doing full download on computer and seeing those changes

Doing a normal sync and removing internet connection. Seeing that the sync in cancelled correctly

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/async/Connection.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5381031&bid=17643949)

timrae commented:
You are brave touching this part of the code... It could open a Pandora's box of problems, or it could be fine, I can't really say either way.

Arthur-Milchior commented:
I don't see how it could be a pandora box. Because either my change is good, as I expect it, and then I'll win more than ten seconds on all sync, (even on sucesful one, where it matters the more) or it'll get quickly caught by an alpha tester and it can be reverted if required (I don't expect it to be the case). I can't imagine how it would add any sneaky bug that does not get detected quickly.

I'm even more confident because upstream does not close the database, and getting closer to upstream is generally a good idea. Especially since I did look at all cases considered in this code and added back the closure each time I wasn't sure that it was safe to keep the code open. 

However, I admit that if it were not for the full publishing process you have, with review by other coder, alpha and beta tester, I would not dare push this PR. Nor any PR actually, because I had never been able to anticipate the few ones which created

timrae commented:
> I can't imagine how it would add any sneaky bug that does not get detected quickly.

Well let's hope so... It would be great to improve the performance here, but the synchronization code path is quite complex and fragile, and I have been bitten in this area several times. It's quite feasible that there are parts of the code that assume that the Collection is closed at the end of a sync which you might not have considered. That's not to say that we shouldn't merge this PR though, I think it's reasonable to try it early on in the next alpha phase, but I do feel it's quite a high risk change that you'll need to keep an eye on.

I would like to say though, there's a reason why we introduced the "How has this been tested" part of the PR template. Of course there are some PRs where it makes sense to skip the template, but I would strongly disagree if you think this PR is such an example. I would like to see detailed explanation of the test scenarios you have run and which code paths they execute before considering merging this.

mikehardy commented:
agreed on testing details and this sync code path being difficult to deal with - I'm open to changes and will read this with a hope to merge it (or any perf improvement) but it's a tough area is all

Mostly responding to say that I am definitely agreed on early in next alpha phase as the earliest I'll get to it though - I'm currently only merging things that are either extremely simple or come packaged with extensive testing until I cut 2.10 which should happen in the next day or three, and then the focus will be on making sure the first couple 2.10 betas work well so there won't be attention to spare for a while

Arthur-Milchior commented:
I added details about the testing in first message. The tests are really simple, that's probably why I didn't spend time about them. Seing whether all kind of sync and the most common failure works. I beg your pardon for making you loose time asking for them

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/async/Connection.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5381031&bid=17643949)

Arthur-Milchior commented:
I want to note that anki@0e5b7da62aa0fd3726ca958f1cdae265d2dc0d91 seems to have changed closing so that it occurs only for full sync

Arthur-Milchior commented:
This has been completely rewritten. I use an enum for conflict resolution, I added comments.
And then I added a variable which state whether or not collection should be closed. It is now closed everytime except when nothing occur and when there is a successful sync

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -1901,17 +1901,17 @@ public void onPostExecute(Payload data) {
                         && ((String) data.data[0]).length() > 0) {
                     // A full sync occurred
-                    String dataString = (String) data.data[0];
+                    Connection.ConflictResolution dataString = (Connection.ConflictResolution) data.data[0];
```
How can `data.data[0] instanceof String` and this line simultaneously be true?

Arthur-Milchior replied:
Nice catch... 
data.data[0] could be either null, "upload" or "download". null meaning it's a normal sync. So I assume the goal was to check whether it was a normal sync.
I removed this enum to put it in another PR, so that both notions gets easier to discuss https://github.com/ankidroid/Anki-Android/pull/7147

mikehardy commented:
Not sure if you want to rebase this after merge of #7150 or not, but they were semi-related. This one didn't take a conflict though, so not necessary, for me to review

mikehardy reviewed (changes requested):
I no longer understand this PR, I think one or more commits have been dropped as the single remaining commit is effectively a no-op now?

Arthur-Milchior commented:
You're right. There is something strang which occured here. Quite probably a rebase error. I'll check later

dae commented:
This can probably be closed, as the new-schema path does not close the collection.

Arthur-Milchior commented:
I'm amazed this only took a single conflict.
Really happy to read this news. Both because it'll save time, and because it mean I won't have to argue for the safety of this PR (which to be honest I had entirely forgotten in the meantime)
